gazebo_plugins/src/gazebo_ros_planar_move.cpp

```diff
@@ -201,20 +285,72 @@ namespace gazebo
     ignition::math::Pose3d pose = parent_->GetWorldPose().Ign();
 #endif
     float yaw = pose.Rot().Yaw();
-    parent_->SetLinearVel(ignition::math::Vector3d(
-          x_ * cosf(yaw) - y_ * sinf(yaw),
-          y_ * cosf(yaw) + x_ * sinf(yaw),
-          0));
-    parent_->SetAngularVel(ignition::math::Vector3d(0, 0, rot_));
+    float dt  = (current_time-this->last_time_).Double();
+
+    ignition::math::Vector3d gravity(parent_->GetWorld()->Gravity());
+    ignition::math::Vector3d linear_vel = ignition::math::Matrix4d(pose.Rot()).Inverse() * parent_->WorldLinearVel();
+    ignition::math::Vector3d angular_vel = parent_->WorldAngularVel();
+
+    // check for joint_state_idel_sec_
+    if ( joint_state_idel_sec_ > 0 &&
+         (current_time-this->last_cmd_subscribe_time_).Double() > joint_state_idel_sec_ ) {
+      if (int((current_time-this->last_cmd_subscribe_time_).Double()*1000)%5000 < 1 ) {
+        ROS_WARN("PlanarMovePlugin (ns = %s) did not received %s for %f sec",
+                 robot_namespace_.c_str(), command_topic_.c_str(), joint_state_idel_sec_);
+      }
+      x_ = y_ = rot_ = 0;
+    }
+    linear_cmd_ = ignition::math::Vector3d (x_, y_, 0);
+    angular_cmd_ = ignition::math::Vector3d(-recover_roll_velocity_p_gain_ * pose.Rot().Roll(),
+                                 -recover_pitch_velocity_p_gain_* pose.Rot().Pitch(),
+                                 rot_);
+    // put on the ground
+    double distBelow;
+    std::string entityName;
+    gazebo::physics::EntityPtr entityBelow;
+    gazebo::physics::EntityPtr fromEntity = parent_;
+    this->GetNearestEntityBelow(fromEntity, distBelow, entityName);
+    if ( ( ! entityName.empty() ) && distBelow >  0.00001 ) {
+      linear_cmd_[2] += recover_z_velocity_p_gain_ * gravity[2] * dt;
+    }
+
+    // accel limit
+    ignition::math::Vector3d linear_acc  = (linear_cmd_  - last_linear_cmd_)/dt;
+    ignition::math::Vector3d angular_acc = (angular_cmd_ - last_angular_cmd_)/dt;
+
+    if ( linear_acc[0] > x_acceleration_limit_max_)  linear_cmd_[0] = last_linear_cmd_[0] + x_acceleration_limit_max_*dt;
```

Comment:
We might write these lines much more compact using an inline `clamp`

Reply:
Ok cool